link-grammar/dict-file/read-dict.c

```diff
@@ -947,14 +954,16 @@ static Exp * make_connector(Dictionary dict)
 			 * Make both a + and a - version, and or them together.  */
 			dict->token[i] = '+';
 			plu = make_dir_connector(dict, i);
+			if (NULL == plu) return NULL;
 			dict->token[i] = '-';
 			min = make_dir_connector(dict, i);
+			if (NULL == min) return NULL;
 
 			n = make_or_node(&dict->exp_list, plu, min);
 		}
 		else
 		{
-			dict_error(dict, "Unknown connector direction type.");
+			dict_error(dict, "Unknown connector direction type '%c'.");
 			return NULL;
```

Comment:
The connector direction is never iven.

Reply:
Indeed.
Since the second argument of dict_error2() is never used, I can change this call to dict_error2(), and change dict_error2() to print an additional value nicely.

A better alternative can be to change dict_error() to a printf-like function.

Reply:
> Um, hang on. condesc_t doesn't have a next pointer in it, so it cannot be chained, so it cannot be used in a hash table, unless you can guarantee zero collisions..... A hash table with zero-or-one items in each bucket ... Hmmm. Are you using some trick to spill to the next empty bucket? There was some old code that did that.

It is similar to the string_set code, in that it uses empty bucket (but in the case of my code I use a constant stride of 1, since it is more cache friendly, see below).  It is also how it is done in the fast-matcher connector lists per word.

This can be improved by using Robin Hood placement (that I use in a WIP to replace the memoizing table of do_count()), and intend to try it here too.

The problem of using a next pointer for chaining is that it is very cache unfriendly, due to random memory access (the chained element has a very low chance to be in a close memory location). On the other hand, searching in an "in place" hash table accesses sequentially a relatively small memory region (very few elements are needed to be inspected by average and the worst case is small, if the table is kept, e.g, under 50% used), and due to the sequential access the CPUs knows to prefetch the next memory to be accessed.